nsv13/code/modules/overmap/ai.dm

```diff
@@ -24,34 +26,71 @@
 		pilot.forceMove(src)
 		gunner = pilot
 	user_thrust_dir = 1
-	if(last_target) //Have we got a target?
+	if(last_target || QDELETED(last_target)) //Have we got a target?
 		if(get_dist(last_target, src) > max_range) //Out of range - Give up the chase
 			last_target = null
 			desired_angle = rand(0,360)
 		else //They're in our range. Calculate a path to them and fire.
 			desired_angle = Get_Angle(src, last_target)
 			fire(last_target) //Fire already handles things like being out of range, so we're good
-	else
-		for(var/X in GLOB.overmap_objects)
-			if(!istype(X, /obj/structure/overmap))
-				continue
-			var/obj/structure/overmap/ship = X
-			if(ship == src || ship.faction == faction || ship.z != z)
-				continue
-			switch(ai_behaviour)
-				if(AI_AGGRESSIVE)
-					if(get_dist(ship,src) <= max_range)
-						target(ship)
-				if(AI_GUARD)
-					if(get_dist(ship,src) <= guard_range)
-						target(ship)
-				if(AI_RETALIATE)
-					if(ship in enemies)
-						target(ship)
+	handle_ai_behaviour()
+
+/obj/structure/overmap/proc/handle_ai_behaviour()
+	if(!ai_controlled)
+		return
+	for(var/X in GLOB.overmap_objects)
+		if(!istype(X, /obj/structure/overmap))
+			continue
+		var/obj/structure/overmap/ship = X
+		if(ship == src || ship.faction == faction || ship.z != z)
+			continue
+		switch(ai_behaviour)
+			if(AI_AGGRESSIVE)
+				if(get_dist(ship,src) <= max_range)
+					target(ship)
+			if(AI_GUARD)
+				if(get_dist(ship,src) <= guard_range)
+					target(ship)
+			if(AI_RETALIATE)
+				if(ship in enemies)
+					target(ship)
+		if(locate(ship in enemies))
+			if(get_dist(ship, src) <= 3)
+				user_thrust_dir = 0 //Don't thrust towards ships we're already close to.
+				brakes = TRUE
+			else
+				user_thrust_dir = 1
+				brakes = FALSE
+
+/obj/structure/overmap/syndicate/ai/carrier/handle_ai_behaviour()
```

Comment:
This is a bad method to do this work out which areas of the proc you can  convert into individual procs to override at that level instead. You dont need to have seperate procs (A even better method is using signals to call the proper behaviour ask @ninjanomnom )

Reply:
You could register on any overmap object that gets added to know when they move rather than using a processing proc to check every tick when they are close